nucliadb_core/src/query_language/mod.rs

```diff
@@ -0,0 +1,442 @@
+
+#![allow(unused)]
+
+use crate::node_error;
+use crate::NodeResult;
+use serde_json::Value as Json;
+use std::collections::HashSet;
+
+pub struct QueryContext {
+    pub paragraph_labels: HashSet<String>,
+    pub field_labels: HashSet<String>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Operator {
+    And,
+    Or,
+    Not,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BooleanOperation {
+    pub operator: Operator,
+    pub operands: Vec<BooleanExpression>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum BooleanExpression {
+    Literal(String),
+    Operation(BooleanOperation),
+}
+
+#[derive(Debug, Clone)]
+struct Translated {
+    is_nested: bool,
+    has_field_labels: bool,
+    has_paragraph_labels: bool,
+    expression: BooleanExpression,
+}
+
+fn translate_literal(literal: String, context: &QueryContext) -> Translated {
+    Translated {
+        is_nested: false,
+        has_field_labels: context.field_labels.contains(&literal),
+        has_paragraph_labels: context.paragraph_labels.contains(&literal),
+        expression: BooleanExpression::Literal(literal),
+    }
+}
+
+fn translate_not(inner: Json, context: &QueryContext) -> NodeResult<Translated> {
+    let inner = translate_json(inner, context)?;
+    Ok(Translated {
+        has_field_labels: inner.has_field_labels,
+        has_paragraph_labels: inner.has_paragraph_labels,
+        is_nested: match &inner.expression {
+            BooleanExpression::Literal(_) => false,
+            BooleanExpression::Operation(_) => true,
+        },
+        expression: BooleanExpression::Operation(BooleanOperation {
+            operator: Operator::Not,
+            operands: vec![inner.expression],
+        }),
+    })
+}
+
+fn translate_other(operator: Operator, operands: Vec<Json>, context: &QueryContext) -> NodeResult<Translated> {
+    let mut is_nested = false;
+    let mut has_field_labels = false;
+    let mut has_paragraph_labels = false;
+    let mut boolean_operation = BooleanOperation {
+        operator,
+        operands: Vec::with_capacity(operands.len()),
+    };
+    for json_operand in operands {
+        let operand = translate_json(json_operand, context)?;
+        if let BooleanExpression::Operation(_) = &operand.expression {
+            is_nested = true;
+        }
+        has_field_labels = has_field_labels || operand.has_field_labels;
+        has_paragraph_labels = has_paragraph_labels || operand.has_paragraph_labels;
+        boolean_operation.operands.push(operand.expression);
+    }
+    Ok(Translated {
+        is_nested,
+        has_field_labels,
+        has_paragraph_labels,
+        expression: BooleanExpression::Operation(boolean_operation),
+    })
+}
+
+fn translate_json(query: Json, context: &QueryContext) -> NodeResult<Translated> {
+    let Json::Object(json_object) = query else {
+        #[rustfmt::skip] return Err(node_error!(
```

Comment:
In my opinion this way error messages pop out in the code, which makes them easier to spot.  Also, even though the message is not that long I think is easier on the eyes and when making changes to have the message as isolated as possible.
I thought, since we have `rustfmt::skip` why not use it in this cases.